File: android/app/src/main/java/com/apoorvdarshan/calorietracker/billing/HostedAIConstants.kt
package com.apoorvdarshan.calorietracker.billing

enum class HostedPlan(val id: String) {
    NONE("none"),
    PLUS("plus"),
    PRO("pro");

    companion object {
        fun fromEntitlements(hasPro: Boolean, hasPlus: Boolean): HostedPlan = when {
            hasPro -> PRO
            hasPlus -> PLUS
            else -> NONE
        }
    }
}

enum class AIMode(val storageValue: String) {
    BYOK("byok"),
    HOSTED("hosted");

    companion object {
        fun fromStorage(value: String?): AIMode =
            entries.firstOrNull { it.storageValue == value } ?: BYOK
    }
}

enum class HostedAIAction(val cost: Int) {
    PHOTO_FOOD(1),
    TEXT_FOOD(1),
    VOICE_FOOD(2),
    INGREDIENT_AI(1),
    REPROCESS_MEAL(1),
    WHAT_IF(1),
    ALLERGENS_LAB(1),
    SIRI_FOOD(1),
    COACH_MESSAGE(1),
    WORKOUT_AI(1),
    HOSTED_STT(1);

    fun manualRecalculateGoals(llmCalls: Int) = llmCalls.coerceAtLeast(1)
}

object HostedAIConstants {
    const val PLUS_ENTITLEMENT = "plus"
    const val PRO_ENTITLEMENT = "pro"

    const val PLUS_MONTHLY = "com.apoorvdarshan.calorietracker.plus.monthly"
    const val PLUS_YEARLY = "com.apoorvdarshan.calorietracker.plus.yearly"
    const val PRO_MONTHLY = "com.apoorvdarshan.calorietracker.pro.monthly"
    const val PRO_YEARLY = "com.apoorvdarshan.calorietracker.pro.yearly"

    const val CREDITS_50 = "com.apoorvdarshan.calorietracker.credits.50"
    const val CREDITS_150 = "com.apoorvdarshan.calorietracker.credits.150"
    const val CREDITS_400 = "com.apoorvdarshan.calorietracker.credits.400"

    const val TIP_SNACK = "com.apoorvdarshan.calorietracker.tip.snack"
    const val TIP_PROTEIN = "com.apoorvdarshan.calorietracker.tip.proteinshake"
    const val TIP_LUNCH = "com.apoorvdarshan.calorietracker.tip.lunch"
    const val TIP_FEAST = "com.apoorvdarshan.calorietracker.tip.feast"

    const val PLUS_DAILY_LIMIT = 30
    const val PRO_DAILY_LIMIT = 60
    const val MAX_HOSTED_IMAGES = 3
    const val HOSTED_AI_BASE_URL = "https://fud-ai.app/api/hosted-ai/v1"
    const val HOSTED_AI_APP_SECRET = "fud-hosted-v1-dev-placeholder"

    val subscriptionProductIds = listOf(PLUS_MONTHLY, PLUS_YEARLY, PRO_MONTHLY, PRO_YEARLY)
    val creditProductIds = listOf(CREDITS_50, CREDITS_150, CREDITS_400)
    val tipProductIds = listOf(TIP_SNACK, TIP_PROTEIN, TIP_LUNCH, TIP_FEAST)

    fun dailyLimit(plan: HostedPlan): Int = when (plan) {
        HostedPlan.PRO -> PRO_DAILY_LIMIT
        HostedPlan.PLUS -> PLUS_DAILY_LIMIT
        HostedPlan.NONE -> 0
    }

    fun creditAmount(productId: String): Int? = when (productId) {
        CREDITS_50 -> 50
        CREDITS_150 -> 150
        CREDITS_400 -> 400
        else -> null
    }
}

sealed class HostedAIQuotaError : Exception() {
    data object NoActiveSubscription : HostedAIQuotaError()
    data class QuotaExceeded(val remainingDaily: Int, val creditBank: Int) : HostedAIQuotaError()
}

File: ios/calorietracker/Models/HostedAIConstants.swift
//
//  HostedAIConstants.swift
//  calorietracker
//
//  Product IDs, entitlements, and hosted AI limits for Fud AI v7.
//

import Foundation

enum HostedAIConstants {
    // MARK: - Entitlements (RevenueCat)

    static let plusEntitlementID = "plus"
    static let proEntitlementID = "pro"

    // MARK: - Subscriptions

    static let plusMonthlyProductID = "com.apoorvdarshan.calorietracker.plus.monthly"
    static let plusYearlyProductID = "com.apoorvdarshan.calorietracker.plus.yearly"
    static let proMonthlyProductID = "com.apoorvdarshan.calorietracker.pro.monthly"
    static let proYearlyProductID = "com.apoorvdarshan.calorietracker.pro.yearly"

    static let subscriptionProductIDs: [String] = [
        plusMonthlyProductID,
        plusYearlyProductID,
        proMonthlyProductID,
        proYearlyProductID,
    ]

    // MARK: - Credit packs (consumables)

    static let credits50ProductID = "com.apoorvdarshan.calorietracker.credits.50"
    static let credits150ProductID = "com.apoorvdarshan.calorietracker.credits.150"
    static let credits400ProductID = "com.apoorvdarshan.calorietracker.credits.400"

    static let creditProductIDs: [String] = [
        credits50ProductID,
        credits150ProductID,
        credits400ProductID,
    ]

    static func creditAmount(for productID: String) -> Int? {
        switch productID {
        case credits50ProductID: 50
        case credits150ProductID: 150
        case credits400ProductID: 400
        default: nil
        }
    }

    // MARK: - Tips (iOS — unchanged; see TipJarView.swift)

    static let tipProductIDs: [String] = [
        "com.apoorvdarshan.calorietracker.tip.snack",
        "com.apoorvdarshan.calorietracker.tip.proteinshake",
        "com.apoorvdarshan.calorietracker.tip.lunch",
        "com.apoorvdarshan.calorietracker.tip.feast",
    ]

    // MARK: - Daily limits

    static let plusDailyLimit = 30
    static let proDailyLimit = 60

    static func dailyLimit(for plan: HostedPlan) -> Int {
        switch plan {
        case .pro: proDailyLimit
        case .plus: plusDailyLimit
        case .none: 0
        }
    }

    // MARK: - Hosted proxy

    /// Production hosted path on fud-ai.app. Override in debug via UserDefaults if needed.
    static let hostedAIBaseURL = "https://fud-ai.app/api/hosted-ai/v1"

    /// The app ships no proxy secret. Requests identify the subscriber with the
    /// RevenueCat app user id (`X-Fud-User-Id`); the Worker verifies the plan
    /// with RevenueCat server-side and meters usage in its own ledger.
    static let hostedUserIDHeader = "X-Fud-User-Id"

    static let maxHostedImages = 3
}

enum HostedPlan: String, Codable, CaseIterable {
    case none
    case plus
    case pro

    var displayName: String {
        switch self {
        case .none: String(localized: "None")
        case .plus: String(localized: "Plus")
        case .pro: String(localized: "Pro")
        }
    }
}

enum AIMode: String, CaseIterable, Identifiable {
    case byok
    case hosted

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .byok: String(localized: "BYOK")
        case .hosted: String(localized: "Hosted (Plus/Pro)")
        }
    }
}

enum HostedAIAction: Equatable {
    case photoFood
    case textFood
    case voiceFood
    case ingredientAI
    case reprocessMeal
    case whatIf
    case allergensLab
    case siriFood
    case coachMessage
    case workoutAI
    case manualRecalculateGoals(llmCalls: Int)
    case hostedSTT

    /// Expected number of hosted round-trips for this action, for UI copy only.
    /// The Worker is the source of truth and meters every upstream call as 1
    /// (so a coach message that needs several tool rounds costs several actions).
    var estimatedCost: Int {
        switch self {
        case .voiceFood: 2
        case .manualRecalculateGoals(let llmCalls): max(1, llmCalls)
        default: 1
        }
    }

    /// Adaptive Goals and onboarding first calc are never metered.
    var isMetered: Bool { true }
}

enum HostedAIQuotaError: LocalizedError, Equatable {
    case notHostedMode
    case noActiveSubscription
    case quotaExceeded(remainingDaily: Int, creditBank: Int)
    case rateLimited

    var errorDescription: String? {
        switch self {
        case .notHostedMode:
            return nil
        case .noActiveSubscription:
            return String(localized: "Subscribe to Plus or Pro to use Hosted AI, or switch to BYOK in Settings → AI Access.")
        case .quotaExceeded:
            return String(localized: "You're out of hosted AI actions for today. Buy credits, upgrade, or switch to BYOK.")
        case .rateLimited:
            return String(localized: "Hosted AI is busy. Please wait a moment and try again.")
        }
    }
}
